### ISSUE TITLE
Why doing loss.backward() and optimizer.step() at different places

### ISSUE THREAD

**Issue by 18445864529** [facebookresearch/fairseq#3985]
The call stack of training is like `train.py - train()` → `trainer.py - train_step()` → `fairseq_task.py - train_step()` → `criterion...`. After the criterion returns the loss to `fairseq_task.py`, the loss is backward there, whereas the optimization step is called in `trainer.py`. 

I'm wondering what's the point of doing these two things in two different files instead of being next to each other?

Because when I return the loss to the outside-most file `train.py`, and do backward() and step() there, I found the log of 'loss' and 'nll_loss' became 'nan', why is this happening? 

Thank you.
